### ISSUE TITLE
ValueError: For early stopping, at least one dataset and eval metric is required for evaluation

### ISSUE THREAD

**Issue by AprLie** [microsoft/LightGBM#3028]
## Environment info

Operating System: Azure Ubuntu16.04

GPU model:v100

Python:3.6.10

LightGBM version or commit hash:2.3.1

## Error message

Traceback (most recent call last):
  File "gbdt2.py", line 48, in <module>
    run()
  File "gbdt2.py", line 35, in run
    early_stopping_rounds=5)
  File "/data/anaconda/envs/torch/lib/python3.6/site-packages/lightgbm/engine.py", line 264, in train
    evaluation_result_list=evaluation_result_list))
  File "/data/anaconda/envs/torch/lib/python3.6/site-packages/lightgbm/callback.py", line 221, in _callback
    _init(env)
  File "/data/anaconda/envs/torch/lib/python3.6/site-packages/lightgbm/callback.py", line 191, in _init
    raise ValueError('For early stopping, '
ValueError: For early stopping, at least one dataset and eval metric is required for evaluation

## Reproducible examples

```python
import lightgbm as lgb
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
import numpy as np


def run():

 
    model = 'gbdt'
    datasplitseed = 42
    tree_seed = 42
    tmp_size = 0.4
    test_size = 0.5
    
    params = { "max_depth":5, 
    "learning_rate":0.001, "bagging_freq":4, "num_iteration":100, "bagging_fraction":0.7, \
    "feature_fraction":0.6, "max_bin":63, "lambda_l1":0, "lambda_l2":0, "min_data_in_leaf":100}
    params['metric'] = 'auc_mu'
    print(params)


    data = np.load(data)
    shape1 = 100
    shape2 = 20
    x_train = np.ones((shape1,shape2))
    y_train = np.ones((shape1))
    x_valid =  np.ones((shape1,shape2))
    y_valid = np.ones((shape1))
    x_test  =np.ones((shape1,shape2))
    y_test = np.ones((shape1))

    lgb_train = lgb.Dataset(x_train,label=y_train)
    lgb_eval = lgb.Dataset(x_valid,label=y_valid, reference=lgb_train)
    gbm = lgb.train(params,
                    lgb_train,
                    num_boost_round=20,
                    valid_sets=[lgb_eval,lgb_train],
                    early_stopping_rounds=5)

    pred = gbm.predict(x_test, num_iteration=gbm.best_iteration)
    target_names = ['IndexName','Index', 'ValueName', 'Aggregation', 'Other']
    print("+ model {}, result:".format(model))
    print(classification_report(y_test, pred, target_names=target_names))
    print("Confusion matrix:")
    print(confusion_matrix(y_test, pred))

if __name__ == '__main__':
    run()
```

**Comment by StrikerRUS** [comment]
Unfortunately, `auc_mu` is not available in `2.3.1` version. It will be included in a next release. Please use another metric or install LightGBM from sources.

**Comment by jlc248** [comment]
Same issue, different metric. Is something as standard as `average_precision` not available in v3.0.0?

OS: CentOS7
Python: v3.7.8
lightgbm: v3.0.0
```
import lightgbm as lgb
import numpy as np

feature_names=['one','two','three','four','five']

train_data = np.random.random((100000,5))
train_labels = np.round(np.random.random(100000)).astype(int)

test_data = np.random.random((20000,5))
test_labels = np.round(np.random.random(20000)).astype(int)
dtest = lgb.Dataset(data=test_data, label=test_labels, feature_name=feature_names)

dtrain = lgb.Dataset(data=train_data, label=train_labels, reference=dtest, feature_name=feature_names)

evals_result = {}

params = {
        'max_depth':-1,          
        'min_child_weight':1e-3, 
        'learning_rate':0.1,     
        'bagging_fraction':1,    
        'feature_fraction':1,  
        'num_leaves':31,      
        'objective':'binary',
        'min_data_in_leaf':20,  
        'lambda_l1':0,          
        'lambda_l2':0,           
        'scale_pos_weight':1,
	'num_threads':20,
	'metric':'average_precision'
    }

model = lgb.train(params,dtrain,500,valid_sets=[dtest], early_stopping_rounds=5,evals_result=evals_result)
```

Error:
`ValueError: For early stopping, at least one dataset and eval metric is required for evaluation`

**Comment by jameslamb** [comment]
Hi @jlc248 ... `average_precision` was added very recently, in #3347 . It will be available in LightGBM 3.1.0, which is in progress (#3484 ).

I believe that all metrics except `average_precision` listed at https://lightgbm.readthedocs.io/en/latest/Parameters.html#metric-parameters are available in LightGBM 3.0.0

**Comment by jlc248** [comment]
Thanks @jameslamb!

**Comment by ankiean** [comment]
I just tried using the average_precision metric, but it was giving me strange results. With early stopping, it says that the first iteration was the very best, but when I print all the validation scores, I see that it is still improving. Basically, it is stopping at the lowest average_precision score. This is wrong, right?

**Comment by jameslamb** [comment]
@ankiean yes, apologies! There was a bug in the Python package that led to the wrong treatment of average precision for early stopping (#3648). This was fixed in #3649 and will be available in the next release.

If you have other bug reports please open a new issue.
